internal/provider/services/jsonDataLoader.go

```diff
@@ -0,0 +1,56 @@
+package services
+
+import (
+	"encoding/json"
+	"log"
+	"os"
+)
+
+// Structs to load data into (from a JSON file for now, will be from DB later)
+type GPGPublicKey struct {
+	KeyID          string `json:"key_id"`
+	ASCIIArmor     string `json:"ascii_armor"`
+	TrustSignature string `json:"trust_signature"`
+	Source         string `json:"source"`
+	SourceURL      string `json:"source_url"`
+}
+
+type SigningKeys struct {
+	GPGPublicKeys []GPGPublicKey `json:"gpg_public_keys"`
+}
+
+type ProviderMetadata struct {
+	OS                  string      `json:"os"`
+	Arch                string      `json:"arch"`
+	Filename            string      `json:"filename"`
+	DownloadURL         string      `json:"download_url"`
+	ShasumsURL          string      `json:"shasums_url"`
+	ShasumsSignatureURL string      `json:"shasums_signature_url"`
+	Shasum              string      `json:"shasum"`
+	SigningKeys         SigningKeys `json:"signing_keys"`
+}
+
+type VersionData struct {
+	Protocols []string           `json:"protocols"`
+	Platforms []ProviderMetadata `json:"platforms"`
+}
+
+type ProviderData map[string]VersionData
+
+var providerObj map[string]ProviderData
+
+func LoadData() (map[string]ProviderData, error) {
```

Comment:
Moving in the right direction...
However we don't want to expose this level of implementation information to the rest of the system.

Add a new Interface here that defines the key operations that you want to do on the ProviderData. 
ie

```go
type ProviderVersionManager interface {
    ListProviderVersions(providerName string) (ProviderData, error)
    GetVersionData(providerName string, providerVersion string, ...) (VersionData, error)
}
```

Then we can add a struct which implements this interface.

```go
type JSONFileProviderVersionManager struct {
    data map[string]ProviderData
}

func NewJSONFileProviderVersionManager() (ProviderVersionManager, error) {
    if data, err := LoadData(); err == nil {
        return &JSONFileProviderVersionManager{ data: data}, nil
    } else {
       return nil, err
    }
}

func (vm *JSONFileProviderVersionManager) ListProviderVersions(providerName string) (ProviderData, error) {}
func (vm *JSONFileProviderVersionManager) GetVersionData(providerName string, providerVersion string, ...) (VersionData, error)
```

Your creation of the provider interface in `cmd\rest_providers_v1.go` would then for the moment pass in the JSONFIleProviderVersionManager.

```go
func runRESTProvidersV1Server(cmd *cobra.Command, args []string) {
        vm, err := services.NewJSONFileProviderVersionManager()
        if err != nil {
            panic(err)
        }
	restAPIServer := providersv1.New(vm)
	startRESTAPIService("rest-providers-v1", mountPathProviders, restAPIServer)
}
```

Then in the rest API you don't have the implementation of how to find the specific data for each request you're just formatting it to return it in the expected format.

Reply:
Updated the code with suggested changes.